### PR TITLE
Add capability to group variables for ReferenceResidualProblem

### PR DESCRIPTION
**Pull request by jiangwen84**
closes Ref #12053

### CODE REVIEW

File: modules/contact/include/ReferenceResidualProblem.h [hunk 1]
```diff
@@ -58,6 +58,13 @@ class ReferenceResidualProblem : public FEProblem
   std::vector<std::string> _refResidVarNames;
   ///@}
 
+  ///@{
+  /// List of grouped solution variable names whose reference residuals will be stored,
+  /// and the residual variable names that will store them.
+  std::vector<std::string> _group_solnVarNames;
```
Comment (bwspenc) [inline modules/contact/include/ReferenceResidualProblem.h]: Follow the MOOSE naming conventions for member variables (something like `_group_soln_var_names`),

Comment (bwspenc) [reply modules/contact/include/ReferenceResidualProblem.h]: It looks like we weren't following those conventions on some of the other variables before (like _solnVarNames). It probably makes sense to go ahead and change those as well.

File: modules/contact/src/ReferenceResidualProblem.C [hunk 2]
```diff
@@ -60,6 +63,49 @@ ReferenceResidualProblem::ReferenceResidualProblem(const InputParameters & param
...
+  std::set<std::string> check_duplicate;
+  if (_group_variables)
+  {
+    for (unsigned int i = 0; i < (*_group_variables).size(); ++i)
+      for (unsigned int j = 0; j < (*_group_variables)[i].size(); ++j)
+        check_duplicate.insert((*_group_variables)[i][j]);
+
+    if (check_duplicate.size() != group_variable_num)
+      mooseError("For group_variables parameter, a variable cannot be grouped with other "
+                 "variables to different groups.");
```
Comment (bwspenc) [inline modules/contact/src/ReferenceResidualProblem.C]: Change this message to:
```
A variable cannot be included in multiple groups in the 'group_variables' parameter

File: modules/contact/src/ReferenceResidualProblem.C [hunk 3]
```diff
@@ -60,6 +63,49 @@ ReferenceResidualProblem::ReferenceResidualProblem(const InputParameters & param
...
+  unsigned int group_variable_num = 0;
+  if (params.isParamValid("group_variables"))
+  {
+    _group_variables = &(params.get<std::vector<std::vector<std::string>>>("group_variables"));
+
+    for (unsigned int i = 0; i < (*_group_variables).size(); ++i)
+    {
+      group_variable_num += (*_group_variables)[i].size();
+      if ((*_group_variables)[i].size() == 1)
+        mooseError("For group_variables parameter, variable ",
```
Comment (bwspenc) [inline modules/contact/src/ReferenceResidualProblem.C]: For group_variables' parameter -> In the 'group_variables parameter

File: modules/contact/src/ReferenceResidualProblem.C [hunk 4]
```diff
@@ -115,6 +161,86 @@ ReferenceResidualProblem::initialSetup()
...
+                                                                                         : false);
+              break;
+            }
+          }
+        }
+      }
+      if (!same_variable)
+        mooseWarning("For group_variables parameter, Row ",
+                     i,
+                     " tries to group standard variables with scalar variables.");
```
Comment (bwspenc) [inline modules/contact/src/ReferenceResidualProblem.C]: "In the 'group_variables' parameter, standard variables and scalar variables are grouped together in group ", i

File: modules/contact/src/ReferenceResidualProblem.C [hunk 5]
```diff
@@ -145,14 +271,30 @@ ReferenceResidualProblem::updateReferenceResidual()
...
+  for (unsigned int i = 0; i < _group_resid.size(); ++i)
+  {
+    _group_resid[i] = 0.0;
+    _group_refResid[i] = 0.0;
+  }
+
   for (unsigned int i = 0; i < _solnVars.size(); ++i)
+  {
     _resid[i] = s.calculate_norm(nonlinear_sys.RHS(), _solnVars[i], DISCRETE_L2);
+    _group_resid[_variable_group_num_index[i]] += pow(_resid[i], 2);
```
Comment (bwspenc) [inline modules/contact/src/ReferenceResidualProblem.C]: Use something like `libMesh::Utility::pow<2>(_resid[i])` everywhere you are raising things to integer powers.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Just some minor stuff.  Overall, I really like this!

File: modules/combined/test/tests/reference_residual/tests [hunk 6]
```diff
@@ -25,4 +25,31 @@
                   Variables/temp/scaling=0.001'
       prereq = test_sm
     [../]
+    [./group_variables]
+      type = 'Exodiff'
+      input = 'group_variables.i'
+      exodiff = 'group_variables_out.e'
+      abs_zero = 1e-9
+      requirement = 'For ReferenceResidualProblem, groups two standard variables and two scalar variables, respectively.'
```
Comment (bwspenc) [inline modules/combined/test/tests/reference_residual/tests]: A little rewording on these requirements:

The reference residual based convergence checking system shall permit the norm of independent groups of either scalar and standard solution variables to be compared against reference quantities provided through the norms of groups of auxiliary variables.

File: modules/combined/test/tests/reference_residual/tests [hunk 7]
```diff
@@ -25,4 +25,31 @@
...
+      design = 'ReferenceResidualProblem.md'
+      issues = '#12053'
+    [../]
+    [./group_variables_duplicate_error_check]
+      type = 'RunException'
+      input = 'group_variables.i'
+      expect_err = "A variable cannot be included in multiple groups in the 'group_variables' parameter."
+      cli_args = 'Problem/group_variables="disp_x disp_y; disp_x scalar_strain_zz1"'
+      requirement = 'Test error for ReferenceResidualProblem, variable disp_x is duplicated in the group_variables.'
```
Comment (bwspenc) [inline modules/combined/test/tests/reference_residual/tests]: The reference residual based convergence checking system shall generate an error if a variable is added to multiple groups.

File: modules/combined/test/tests/reference_residual/tests [hunk 8]
```diff
@@ -25,4 +25,31 @@
...
+      design = 'ReferenceResidualProblem.md'
+      issues = '#12053'
+    [../]
+    [./group_variables_ungroup_error_check]
+      type = 'RunException'
+      input = 'group_variables.i'
+      expect_err = "In the 'group_variables' parameter, variable disp_x is not grouped with other variables."
+      cli_args = 'Problem/group_variables="disp_x; scalar_strain_zz1 scalar_strain_zz2"'
+      requirement = 'Test error for ReferenceResidualProblem, variable disp_x is not grouped with other variable.'
```
Comment (bwspenc) [inline modules/combined/test/tests/reference_residual/tests]: The reference residual based convergence checking system shall generate an error if a group containing a single variable is defined.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Almost there.  Just change the wording on the requirements as noted.

**Review by bwspenc — APPROVED** [review]

**Comment by jiangwen84** [comment]
ReferenceResidualProblem will not improve the convergence but just give a better way (for some cases) to check the convergence. This PR just adds a new capability to group several variables together, such as group scalar variables or all the displacements.

**Pull request merged**